diff -urN a/rust/private/rustc.bzl b/rust/private/rustc.bzl
--- a/rust/private/rustc.bzl	2025-12-09 17:08:33.000000000 -0800
+++ b/rust/private/rustc.bzl	2026-02-27 10:57:19.934593354 -0800
@@ -1479,8 +1479,9 @@
     if toolchain.target_os == "windows" and crate_info.type == "cdylib":
         # Rustc generates the import library with a `.dll.lib` extension rather than the usual `.lib` one that msvc
         # expects (see https://github.com/rust-lang/rust/pull/29520 for more context).
-        interface_library = ctx.actions.declare_file(crate_info.output.basename + ".lib", sibling = crate_info.output)
-        outputs.append(interface_library)
+        if not experimental_use_cc_common_link:
+            interface_library = ctx.actions.declare_file(crate_info.output.basename.removesuffix(".dll").removesuffix(".DLL") + ".dll.lib", sibling = crate_info.output)
+            outputs.append(interface_library)

     # The action might generate extra output that we don't want to include in the `DefaultInfo` files.
     action_outputs = list(outputs)
@@ -1630,7 +1631,7 @@
             pdb_file = ctx.actions.declare_file(crate_info.output.basename[:-len(crate_info.output.extension)] + "pdb", sibling = crate_info.output)
             additional_linker_outputs.append(pdb_file)

-        cc_common.link(
+        linking_outputs = cc_common.link(
             actions = ctx.actions,
             feature_configuration = feature_configuration,
             cc_toolchain = cc_toolchain,
@@ -1642,6 +1643,10 @@
             additional_outputs = additional_linker_outputs,
         )

+        if toolchain.target_os == "windows" and crate_info.type == "cdylib":
+            if linking_outputs.library_to_link and linking_outputs.library_to_link.interface_library:
+                interface_library = linking_outputs.library_to_link.interface_library
+
         outputs = [crate_info.output]

     coverage_runfiles = []
@@ -1736,8 +1741,6 @@
     else:
         providers.extend([crate_info, dep_info])

-    providers += establish_cc_info(ctx, attr, crate_info, toolchain, cc_toolchain, feature_configuration, interface_library)
-
     output_group_info = {}

     if pdb_file:
@@ -1754,6 +1757,8 @@
     if output_group_info:
         providers.append(OutputGroupInfo(**output_group_info))

+    providers += establish_cc_info(ctx, attr, crate_info, toolchain, cc_toolchain, feature_configuration, interface_library)
+
     # A bit unfortunate, but sidecar the lints info so rustdoc can access the
     # set of lints from the target it is documenting.
     if hasattr(ctx.attr, "lint_config") and ctx.attr.lint_config:
@@ -1972,7 +1977,7 @@
             else:
                 cc_infos.append(dep.cc_info)

-    if crate_info.type in ("rlib", "lib"):
+    if crate_info.type in ("rlib", "lib") or (toolchain.target_os == "windows" and crate_info.type in ("cdylib", "staticlib")):
         libstd_and_allocator_cc_info = _get_std_and_alloc_info(ctx, toolchain, crate_info)
         if libstd_and_allocator_cc_info:
             # TODO: if we already have an rlib in our deps, we could skip this
