From 6c49377b75bb924540623b2d0356864414f534d4 Mon Sep 17 00:00:00 2001
From: Will Hollins <whollins@google.com>
Date: Mon, 22 Sep 2025 13:49:15 +0100
Subject: [PATCH] GRPC patch to ares_build.h

---
 include/ares_build.h | 20 ++++++++++++++++++--
 1 file changed, 18 insertions(+), 2 deletions(-)

diff --git a/include/ares_build.h b/include/ares_build.h
index 7f7d2e02..e5988cfe 100644
--- a/include/ares_build.h
+++ b/include/ares_build.h
@@ -192,8 +192,13 @@
 #endif
 
 /* Data type definition of ares_ssize_t. */
+/* gRPC Manuel edit here!
+ * Possibly include <_mingw.h> header to define __int64 type under mingw */
 #ifdef _WIN32
 #  ifdef _WIN64
+#    ifdef __MINGW32__
+#      include <_mingw.h>
+#    endif
 #    define CARES_TYPEOF_ARES_SSIZE_T __int64
 #  else
 #    define CARES_TYPEOF_ARES_SSIZE_T long
@@ -204,4 +209,15 @@
 
 typedef CARES_TYPEOF_ARES_SSIZE_T ares_ssize_t;
 
+/* IMPORTANT: gRPC MANUAL EDIT HERE!
+ * Undefine UNICODE, as c-ares does not use the ANSI version of functions
+ * explicitly. */
+#ifdef UNICODE
+#  undef UNICODE
+#endif
+
+#ifdef _UNICODE
+#  undef _UNICODE
+#endif
+
 #endif /* __CARES_BUILD_H */
-- 
2.51.0.470.ga7dc726c21-goog

