From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Zach Yu <zachyu@google.com>
Date: Mon, 6 May 2024 16:19:42 -0700
Subject: Fix cc_common_link when using sibling repository layout.


diff --git a/rust/private/rustc.bzl b/rust/private/rustc.bzl
index 474d1a61..579cb682 100644
--- a/rust/private/rustc.bzl
+++ b/rust/private/rustc.bzl
@@ -1391,13 +1391,17 @@ def rustc_compile_action(
 
         # The path to the package dir, including a trailing "/".
         package_dir = ctx.bin_dir.path + "/"
-        if ctx.label.workspace_root:
+        # For external repositories, workspace root is not part of the output
+        # path when sibling repository layout is used (the repository name is
+        # part of the bin_dir). This scenario happens when the workspace root
+        # starts with "../"
+        if ctx.label.workspace_root and not ctx.label.workspace_root.startswith("../"):
             package_dir = package_dir + ctx.label.workspace_root + "/"
         if ctx.label.package:
             package_dir = package_dir + ctx.label.package + "/"
 
         if not crate_info.output.path.startswith(package_dir):
-            fail("The package dir path {} should be a prefix of the crate_info.output.path {}", package_dir, crate_info.output.path)
+            fail("The package dir path", package_dir, "should be a prefix of the crate_info.output.path", crate_info.output.path)
 
         output_relative_to_package = crate_info.output.path[len(package_dir):]
 
-- 
2.45.0.rc1.225.g2a3ae87e7f-goog

