--- a/tink/internal/ssl_unique_ptr.h
+++ b/tink/internal/ssl_unique_ptr.h
@@ -23,14 +23,12 @@
 // definition of OPENSSL_IS_BORINGSSL in case BoringSSL is used.
 #include "openssl/crypto.h"
 
-#ifndef OPENSSL_IS_BORINGSSL
 #include "openssl/bn.h"
 #include "openssl/cmac.h"
 #include "openssl/ec.h"
 #include "openssl/evp.h"
 #include "openssl/hmac.h"
 #include "openssl/rsa.h"
-#endif
 
 namespace crypto {
 namespace tink {
