--- a/tink/internal/ssl_unique_ptr.h
+++ b/tink/internal/ssl_unique_ptr.h
@@ -23,14 +23,12 @@
 // definition of OPENSSL_IS_BORINGSSL in case BoringSSL is used.
 #include "openssl/crypto.h"
 
-#ifndef OPENSSL_IS_BORINGSSL
 #include "openssl/bn.h"
 #include "openssl/cmac.h"
 #include "openssl/ec.h"
 #include "openssl/evp.h"
 #include "openssl/hmac.h"
 #include "openssl/rsa.h"
-#endif
 
 namespace crypto {
 namespace tink {
--- a/tink/subtle/decrypting_random_access_stream.cc
+++ b/tink/subtle/decrypting_random_access_stream.cc
@@ -58,7 +58,7 @@
   }
   std::unique_ptr<DecryptingRandomAccessStream> dec_stream(
       new DecryptingRandomAccessStream());
-  absl::MutexLock lock(dec_stream->status_mutex_);
+  absl::MutexLock lock(&dec_stream->status_mutex_);
   dec_stream->segment_decrypter_ = std::move(segment_decrypter);
   dec_stream->ct_source_ = std::move(ciphertext_source);
 
@@ -100,7 +100,7 @@
 absl::Status DecryptingRandomAccessStream::PRead(
     int64_t offset, int size, util::SecretData* buffer) {
   {
-    absl::MutexLock lock(status_mutex_);
+    absl::MutexLock lock(&status_mutex_);
     if (!status_.ok()) return status_;
   }
   if (offset < 0) return absl::Status(absl::StatusCode::kInvalidArgument,
@@ -300,7 +300,7 @@
 
 absl::StatusOr<int64_t> DecryptingRandomAccessStream::size() {
   {  // Initialize, if not initialized yet.
-    absl::MutexLock lock(status_mutex_);
+    absl::MutexLock lock(&status_mutex_);
     InitializeIfNeeded();
     if (!status_.ok()) return status_;
   }
